Finalize deleting user associations

lightalloy opened the pull request:
## What type of PR is this? (check all applicable)
- [x] Bug Fix
- [x] Added Specs

## Description
- added tests to check that all of the user associations are deleted when a user is deleted via `Users::Delete`
To be able to do this I had to:
- alter factories so that they would create valid records
- define some of the user associations appropriately
- add `dependent` option on some of the user associations
- add explicit associations deletion to `Users::DeleteActivity` since `delete` that we use in `Users::Delete` doesn't delete dependent associations

Initially, I planned to add tests for `Users::DeleteActivity` but not all association records are deleted in this class, so I decided that it makes more sense to add a test for the user deletion in general.
While working on this test I found out possible problems that users could face while self-deleting their account (https://github.com/thepracticaldev/tech-private/issues/343). Some of the associated records were not deleted before `user.delete`, which could cause foreign key violation error, and so, 500 error for a user.

## Related Tickets & Documents
 #4795

lightalloy commented on app/models/user.rb:
```diff
@@ -21,34 +21,39 @@ class User < ApplicationRecord
   has_many :collections, dependent: :destroy
   has_many :comments, dependent: :destroy
-  has_many :email_messages, class_name: "Ahoy::Message"
+  has_many :email_messages, class_name: "Ahoy::Message", dependent: :destroy
```
I had to specify `dependent` options because otherwise the associated records wouldn't be deleted via `delete_all` (default strategy is `nullify`)
And it doesn't make sense to keep orphaned records in most cases.

lightalloy commented on app/models/user.rb:
```diff
@@ -21,34 +21,39 @@ class User < ApplicationRecord
-  has_many :feedback_messages
-  has_many :rating_votes
+
+  has_many :offender_feedback_messages, class_name: "FeedbackMessage", inverse_of: :offender, foreign_key: :offender_id, dependent: :delete_all
```
Specified multiple associations instead of one `feedback_messages` w/o options which failed because there is no `user_id` in the `feedback_messages` table.

lightalloy commented on app/models/user.rb:
```diff
@@ -21,34 +21,39 @@ class User < ApplicationRecord
   has_many :access_tokens, class_name: "Doorkeeper::AccessToken", foreign_key: :resource_owner_id, inverse_of: :resource_owner, dependent: :delete_all
   has_many :webhook_endpoints, class_name: "Webhook::Endpoint", foreign_key: :user_id, inverse_of: :user, dependent: :delete_all
-  has_many :user_blocks
+  has_many :blocker_blocks, class_name: "UserBlock", foreign_key: :blocker_id, inverse_of: :blocker, dependent: :delete_all
```
Same as with `feedback_messages`, no `user_blocks.user_id`

lightalloy commented on app/models/feedback_message.rb:
```diff
@@ -1,8 +1,7 @@
-  belongs_to :reviewer, foreign_key: "reviewer_id", class_name: "User", optional: true, inverse_of: :feedback_messages
-  belongs_to :reporter, foreign_key: "reporter_id", class_name: "User", optional: true, inverse_of: :feedback_messages
-  belongs_to :affected, foreign_key: "affected_id", class_name: "User", optional: true, inverse_of: :feedback_messages
+  belongs_to :offender, foreign_key: "offender_id", class_name: "User", optional: true, inverse_of: :offender_feedback_messages
```
Changed `inverse_of` options so that the associations would work properly.

Zhao-Andy replied:
Ahh thanks, I could never figure this out 😓

lightalloy commented on app/models/feedback_message.rb:
```diff
@@ -1,8 +1,7 @@
 class FeedbackMessage < ApplicationRecord
-  belongs_to :offender, foreign_key: "offender_id", class_name: "User", optional: true, inverse_of: :feedback_messages
-  belongs_to :reviewer, foreign_key: "reviewer_id", class_name: "User", optional: true, inverse_of: :feedback_messages
```
Removed `reviewer` association because there is no `reviewer_id` in the `feedback_messages` table, so the association seems to be unused.

lightalloy commented on app/models/user_block.rb:
```diff
@@ -1,6 +1,6 @@
 class UserBlock < ApplicationRecord
-  belongs_to :blocker, foreign_key: "blocker_id", class_name: "User", inverse_of: :user_blocks
-  belongs_to :blocked, foreign_key: "blocked_id", class_name: "User", inverse_of: :user_blocks
+  belongs_to :blocker, foreign_key: "blocker_id", class_name: "User", inverse_of: :blocker_blocks
```
Changed `inverse_of` options so that the associations would work properly.

lightalloy commented on app/services/users/delete_activity.rb:
```diff
@@ -12,6 +12,34 @@ def call(user)
+      user.rating_votes.delete_all
+      user.tweets.delete_all
+
+      user.offender_feedback_messages.delete_all
```
Not sure how `feedback_messages` are used currently. Please, tell me if we need to keep them.

Zhao-Andy replied:
These are essentially the abuse reports that someone makes. I think it would be safer to remove the lines, but it's not a big deal since I don't think any `feedback_messages` are actually associated to any user currently.

lightalloy commented on spec/services/users/delete_spec.rb:
```diff
@@ -34,4 +34,37 @@
+      user.reload
+      described_class.call(user)
+      user_associations.each do |user_association|
+        expect { user_association.reload }.to raise_error(ActiveRecord::RecordNotFound)
```
This test has a lot of expectations which is not very convenient, but I chose this over calling `Users::Delete` multiple times. Please, share if you have better ideas.

rhymes replied:
Makes sense! 

you can have

`    it "deletes all the associations" do, :aggregate_failures`

so that it will aggregate all failures

citizen428 replied:
Minor typo, should be `it "deletes all the associations", :aggregate_failures do`

citizen428 replied:
I also quite like the [block form](https://relishapp.com/rspec/rspec-core/docs/expectation-framework-integration/aggregating-failures#use-%60aggregate-failures%60-block-form), maybe we can use it to make some smaller aggregate groups, split by topic or something.

lightalloy commented on app/services/users/delete_activity.rb:
```diff
@@ -12,6 +12,33 @@ def call(user)
       user.badge_achievements.delete_all
       user.github_repos.delete_all
+
+      user.access_grants.delete_all
```
I had to add a lot of deletions to `DeleteActivity` because otherwise, they will keep hanging out in the database.

rhymes commented on spec/factories/comments.rb:
```diff
@@ -1,11 +1,7 @@
-      association :commentable, factory: :article
-    end
+    body_markdown { Faker::Hipster.paragraph(sentence_count: 1) }
+    association :commentable, factory: :article
```
I like this, though I wonder if this doesn't increase the amount of objects created per each test that doesn't explicitly call `create(:comment, commentable: ...)`

lightalloy replied:
If `commentable` (or both `commentable_id` and `commentable_type`) was not specified before, creating comment would fail because with a validation error `Commentable must exist`. What was happening before the change: `commentable_id` was specified in the tests when using a factory and the default value was used for the `commentable_type` field. So, the associated records were created anyway.
Now the factory can be used both with or without specifying `commentable`.

citizen428 commented on app/models/user.rb:
```diff
@@ -21,34 +21,39 @@ class User < ApplicationRecord
   has_many :notes, as: :noteable, inverse_of: :noteable
-  has_many :profile_pins, as: :profile, inverse_of: :profile
-  has_many :authored_notes, as: :author, inverse_of: :author, class_name: "Note"
+  has_many :profile_pins, as: :profile, inverse_of: :profile, dependent: :delete_all
```
Maybe because it's faster, since `delete_all` directly deletes from the DB. Makes sense since there are no other associations on `ProfilePin` we'd want to trigger here via `:destroy`.

lightalloy replied:
Yep, that's for faster deletion. In this case, I `delete_all` will be called anyway because I call it explicitly in `Users::DeleteActivity`, so that shouldn't affect the logic. I don't have a strong opinion if `delete_all` or `destroy` is needed here. As I know, all user deletions should be done via `Users::Delete`, not directly.

rhymes replied:
makes sense! thank you for the explanation

citizen428 commented on app/services/users/delete_activity.rb:
```diff
@@ -12,6 +12,33 @@ def call(user)
+      user.created_podcasts.update_all(creator_id: nil)
+
+      user.blocker_blocks.delete_all
+      user.blocked_blocks. delete_all
```
An extra space snuck in here before `delete_all` 😃

lightalloy commented:
As for the `DeleteActivity.call`, I'm not happy with its length, but I don't see clear associations groups to divide it in a way so that it would help with understanding. So I think it's not worth to split it for the CodeClimate happiness. If you have ideas for refactoring, that could be an opportunity for the future pr.

lightalloy commented:
@rhymes @citizen428 Thanks for your reviews! :rocket:

lightalloy commented:
I changed the `feedback_messages` `dependent` strategy to `:nullify` so that they won't be deleted with `delete_all`. I have also added a test to check that `kept associations` are not deleted.
